### PR TITLE
Port std::map to gRPC Core

### PR DESCRIPTION
**Pull request by mhaidrygoog**
Addresses Issue [17864](https://github.com/grpc/grpc/issues/17864).

### CODE REVIEW

**Comment by mhaidrygoog** [comment]
Yes I'm planning to use this for locality map and another exercise may be to refactor other code like subchannel pool to use this instead of AVL map

File: src/core/lib/gprpp/memory.h [hunk 1]
```diff
@@ -71,7 +71,7 @@ inline UniquePtr<T> MakeUnique(Args&&... args) {
 
 // an allocator that uses gpr_malloc/gpr_free
 template <class T>
-class Allocator {
+class Allocator : public std::allocator<T> {
```
Comment (markdroth) [inline src/core/lib/gprpp/memory.h]: Why is this needed?  This is only used in templates, so as long as `Allocator` provides the same API as `std::allocator`, it should not need to actually inherit from it.

File: test/core/gprpp/map_test.cc [hunk 2]
```diff
@@ -0,0 +1,99 @@
...
+
+#include "src/core/lib/gprpp/map.h"
+#include <gtest/gtest.h>
+#include "src/core/lib/gprpp/memory.h"
+#include "test/core/util/test_config.h"
+
+namespace grpc_core {
+namespace testing {
+TEST(MapTest, MapAddEntriesTest) {
+  grpc_core::Map<char, int> map;
```
Comment (markdroth) [inline test/core/gprpp/map_test.cc]: No need for `grpc_core::` here, since we're already inside of that namespace.

Same thing throughout.

**Comment by mhaidrygoog** [comment]
Sounds good. I'll wrap my changes in the xDs Locality Map PR.

File: test/core/gprpp/map_test.cc (hunk 2 above)
Comment (mhaidrygoog) [reply test/core/gprpp/map_test.cc]: Sounds good, will make the change in the xDS PR

File: src/core/lib/gprpp/memory.h (hunk 1 above)
Comment (mhaidrygoog) [reply src/core/lib/gprpp/memory.h]: Your right, I assumed it must inherit to be a type of allocator but the code worked without the inheritance.
